nomad/state/state_store.go

```diff
@@ -5760,9 +5760,27 @@ func (s *StateStore) ScalingPoliciesByNamespace(ws memdb.WatchSet, namespace, ty
 	return iter, nil
 }
 
-func (s *StateStore) ScalingPoliciesByJob(ws memdb.WatchSet, namespace, jobID string) (memdb.ResultIterator, error) {
+func (s *StateStore) ScalingPoliciesByJob(ws memdb.WatchSet, namespace, jobID, policyType string) (memdb.ResultIterator,
+	error) {
 	txn := s.db.ReadTxn()
-	return s.ScalingPoliciesByJobTxn(ws, namespace, jobID, txn)
+	iter, err := s.ScalingPoliciesByJobTxn(ws, namespace, jobID, txn)
+	if err != nil {
+		return nil, err
+	}
+
+	if policyType == "" {
+		return iter, nil
+	}
+
+	filter := func(raw interface{}) bool {
+		p, ok := raw.(*structs.ScalingPolicy)
+		if !ok {
+			return true
+		}
+		return policyType != p.Type
```

Comment:
I think my only comment about this is that it seems like this check could be moved into `ScalingPoliciesByJobTxn`, where the new `policyType string` would be passed through to it.

Looking at `ScalingPoliciesByJobTxn`, it already does the same kind of type assertion we see here, so it happening again (specifically the `!ok` part, which I think is _inconceivable_) makes me want a single `filter` that lives in a single place.

Reply:
yeah, maybe 50/50? this short-circuits the filter on the outside of the loop when `policyType == ""`, as opposed to doing it on every filter operation. merging them avoids duplicate cast assertion checks.